vaadin-menu-bar-flow-parent/vaadin-menu-bar-flow/src/main/java/com/vaadin/flow/component/menubar/MenuBar.java

```diff
@@ -254,6 +262,60 @@ public void removeThemeVariants(MenuBarVariant... variants) {
                         .collect(Collectors.toList()));
     }
 
+    /**
+     * Gets the internationalization object previously set for this component.
+     * <p>
+     * Note: updating the object content that is gotten from this method will
+     * not update the lang on the component if not set back using
+     * {@link MenuBar#setI18n(MenuBarI18n)}
+     *
+     * @return the i18n object. It will be <code>null</code>, If the i18n
+     *         properties weren't set.
+     */
+    public MenuBarI18n getI18n() {
+        return i18n;
+    }
+
+    /**
+     * Sets the internationalization properties for this component.
+     *
+     * @param i18n
+     *            the internationalized properties, not <code>null</code>
+     */
+    public void setI18n(MenuBarI18n i18n) {
+        Objects.requireNonNull(i18n,
+                "The I18N properties object should not be null");
+        this.i18n = i18n;
+
+        runBeforeClientResponse(ui -> {
+            if (i18n == this.i18n) {
+                setI18nWithJS();
+            }
+        });
+    }
+
+    private void setI18nWithJS() {
```

Comment:
I think this also needs to be called when attaching the component (`onAttach`). I recently learned that changes made with `executeJs` do not persist when detaching and reattaching the component. We could verify this by adding an IT that detaches and reattaches the component, and check that the custom translation is still applied.

Also when calling this from onAttach, the I18N object might be null, which would cause NPEs.

Reply:
You were right, fixed!